File: csharp/minesweeper/Minesweeper.cs
using System;
using System.Linq;

public static class Minesweeper
{
    public static string[] Annotate(string[] input)
    {
        var mineCounts =
            input.SelectMany((line, y) => 
                line
                .Select((c, x) => (isMine: c == '*', x, y))
                .Where(cell => cell.isMine))
            .SelectMany(cell => 
                Enumerable.Range(cell.x - 1, 3).SelectMany(x => 
                Enumerable.Range(cell.y - 1, 3).Where(y =>
                    x != cell.x || y != cell.y)
                .Select(y => (x, y))))
            .GroupBy(cell => cell)
            .ToDictionary(cell => (cell.Key.x, cell.Key.y), cell => cell.Count());
        
        return input.Select((line, y) => 
            string.Concat(line.Select((c, x) => 
                c == '*' ? "*"
                : (mineCounts.ContainsKey((x, y)) 
                    ? mineCounts[(x, y)].ToString() 
                    : " "))))
            .ToArray();
    }
}

File: fsharp/minesweeper/Minesweeper.fs
module Minesweeper
open System

let annotate (input: string list) = 
    let minesAndAdjancies = 
        let adjacencies point =
            [-1..1] |> List.collect (fun y -> [-1..1] |> List.map (fun x -> (fst point + y, snd point + x)))
            |> List.except [point]
        let forPoint y x c = if c = '*' then ((y,x), adjacencies (y,x) |> Some) else (((y,x)), None)
        input 
            |> List.mapi (fun y line -> line |> Seq.toList |> List.mapi (fun x c -> forPoint y x c))
            |> List.concat |> List.filter (fun (_,a) -> a.IsSome) |> List.map (fun (p,a) -> (p,a.Value))
    let plotter y x c = 
        if c = '*' then '*' else 
            let count = List.filter (fun (_,a) -> List.contains (y,x) a) minesAndAdjancies |> List.length
            if count = 0 then ' ' else count |> string |> char
    input 
        |> List.mapi (fun y line -> line |> Seq.toList |> List.mapi (fun x c -> plotter y x c) |> String.Concat)